### ISSUE TITLE
With multithreading, only one process is using CPU

### ISSUE THREAD

**Issue by jmeppley** [gjospin/PhyloSift#457]
When I run phylosift with multiple threads, in the lastal and pplacer steps, only one process is using the CPU. In top I see multiple lastal processes (one for each thread requested plus one more). However, only one lastal process will show a usage of close to 100% and the rest will be in the single digits. 

```
Processes: 76 total, 7 running, 69 sleeping, 235 threads               12:42:32
Load Avg: 1.97, 6.11, 4.28  CPU usage: 13.22% user, 0.36% sys, 86.40% idle
SharedLibs: 2968K resident, 1392K data, 0B linkedit.
MemRegions: 6029 total, 529M resident, 10M private, 162M shared.
PhysMem: 1703M wired, 359M active, 10G inactive, 12G used, 12G free.
VM: 183G vsize, 1035M framework vsize, 2857174(0) pageins, 204426(0) pageouts.
Networks: packets: 21906471/20G in, 22404064/13G out.
Disks: 71257/3593M read, 560910/24G written.

PID    COMMAND      %CPU TIME     #TH  #WQ  #POR #MRE RPRVT  RSHRD  RSIZE  VPRVT
15687  lastal       99.3 01:05.82 1/1  0    14   34   315M   492K   315M   389M
15677  lastal       4.1  00:01.05 1    0    14   31   660K   92M    93M    74M
15680  perl5.10.0   1.4  00:00.88 1    0    8    107  15M+   34M    18M+   18M
15579  perl5.10.0   0.9  00:01.48 1    0    18   102  732K   34M    34M    732K
15659  gzip         0.0  00:00.02 1    0    14   23   232K   296K   492K   74M
15682  lastal       0.0  00:01.01 1    0    14   31   764K   92M    93M    74M
92851  sshd         0.0  00:00.52 1    0    11   70   300K   3680K  756K   536K
15684  lastal       0.0  00:01.12 1    0    14   31   768K   92M    93M    74M
92855  bash         0.0  00:00.02 1    0    17   24   432K   760K   1096K  73M
92731  bash         0.0  00:00.18 1    0    17   24   620K   760K   1304K  73M
92728  sshd         0.0  00:01.08 1    0    11   74   344K   3764K  756K   580K
92726  launchd      0.0  00:01.98 2    0    43   53   460K   428K   852K   60M
15686  lastal       0.0  00:01.05 1    0    14   31   756K   92M    93M    74M
15685  lastal       0.0  00:01.00 1    0    14   31   660K   92M    93M    74M
15683  lastal       0.0  00:00.91 1    0    14   31   656K   92M    93M    74M
15681  lastal       0.0  00:01.07 1    0    14   31   660K   92M    93M    74M
15679  lastal       0.0  00:00.95 1    0    14   31   660K   92M    93M    74M
```

The same happens when pplacer runs.

I don't think memory is a problem. I have tried on a machine with 24GB and one with 300GB. Disk access shouldn't be a problem. Both systems use a striped raid-like file system that usually is really fast.

Running lastal with simple file fragmenting I get full usage:

```
Processes: 93 total, 10 running, 83 sleeping, 258 threads              12:36:42
Load Avg: 8.13, 5.98, 2.81  CPU usage: 99.29% user, 0.70% sys, 0.0% idle
SharedLibs: 2968K resident, 1392K data, 0B linkedit.
MemRegions: 6223 total, 2328M resident, 11M private, 3549M shared.
PhysMem: 1758M wired, 5630M active, 6429M inactive, 13G used, 11G free.
VM: 251G vsize, 1035M framework vsize, 2648714(0) pageins, 204426(0) pageouts.
Networks: packets: 21264879/19G in, 21881131/13G out.
Disks: 70866/3548M read, 559255/24G written.

PID    COMMAND      %CPU   TIME     #TH  #WQ  #POR #MRE RPRVT  RSHRD  RSIZE
15450  lastal       98.8   03:38.17 1/1  0    14   46+  265M+  744M   3674M+
15430  lastal       98.8   03:40.15 1/1  0    14   43   265M+  744M   3674M+
15459  lastal       98.5   03:35.84 1/1  0    14   43   265M   744M   3673M
15454  lastal       98.5   03:39.87 1/1  0    14   43   265M+  744M   3673M+
15445  lastal       98.1   03:41.47 1/1  0    14   43   265M   744M   3673M
15440  lastal       97.7   03:37.63 1/1  0    14   44   265M   744M   3674M
15464  lastal       97.1   03:39.14 1/1  0    14   44   265M   744M   3674M
```

Is there some bottleneck I can work around?

**Comment by koadman** [comment]
Hi John, would you happen to be running 16S amplicon data through phylosift? The n+1 `lastal` processes you observe include n processes to search the protein database and 1 process to search the rRNA database. This parallelization strategy works well for metagenomic data but obviously falls apart when using 16S amplicon data. As for `pplacer`, phylosift passes along the requested number of threads to `pplacer`, which I believe begins with a serial component before forking processes to run in parallel. The serial component can be nontrivial when placing lots of reads (e.g. amplicon data).

I've been working to improve phylosift's behavior on amplicon data in the past week as I've had a few such datasets come in recently to analyze myself and will push these changes in since it sounds like it may help you.

**Comment by jmeppley** [comment]
No, this is happening on metagenomic data. But that does explain the extra thread.

**Comment by jmeppley** [comment]
OK, I see pplacer going full on all threads now. It did just take a while. I'll play around more with last and see if I can't figure out what my problem is...

**Comment by jmeppley** [comment]
It's not multithreading. The protein search is just limited by CPU. With one thread, I get this:

```
PID    COMMAND      100.0     T3:25:10 1/1  #WQ  #POR #MREGS RPRV   RSHRD
49027  lastal       99.7      00:40.90 1/1  0    14   34     314M   492K
49026  lastal       11.2      00:04.76 1/1  0    14   31     93M    492K
```

That still doesn't seem right, though.

**Comment by jmeppley** [comment]
Does octopus hand out threads to the nucleotide and protein searches in sync? How hard would it be to put the multithreading on the nucleotide side? It looks like search is taking about half the total time for my runs.

**Comment by koadman** [comment]
I was able to take a closer look at this over the weekend, and indeed the RNA search is bottlenecking even on metagenomic data. There is some amount of synchronicity imposed on the RNA & protein searches due to the way the octopus works: a set of child processes are spawned, one for each lastal search, and then the parent process feeds sequences to the children. The RNA search child receives all of the sequences, while the sequences are divided among the 1 or more protein search children. When the RNA search is slower than the protein searches, the input buffer to the RNA lastal process fills and the parent process waits. It does not continue feeding sequences to the other processes.

So, the change you suggested of parallelizing the RNA search should resolve the issue. It looks like this problem emerged when we enabled FastQ quality-aware RNA searches with lastal some time ago. lastal processes the same sequences in FastA format (e.g. without qualities) far more quickly. So a temporary workaround would be to convert your FastQ file to FastA before using it with phylosift. 
We haven't evaluated carefully whether using lastal's quality-aware alignment option improves search accuracy at all. My guess is that the impact on results from the overall phylosift pipeline is negligible, especially considering that phylosift itself imposes quality trimming prior to the lastal search. However, the impact on runtime is clearly large and so it might make sense for us to disable the quality-aware lastal alignment in an upcoming release.
